File: 卡码/代码随想录/1/07-回溯/C++/02-分割/LeetCode_0131.cpp
//131. 分割回文串
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
	vector<string> path;
	vector<vector<string>> res;
public:
	bool isPalindrome(string s,int start,int end){
		while(start<end){
			if(s[start]!=s[end]) return false;
			start++;
			end--;
		}
		return true;
	}
	void backtracking(string s,int index){
		if(index>=s.length()){
			res.push_back(vector<string>(path));
			return;
		}
		for(int i=index;i<s.length();i++){
			if(isPalindrome(s,index,i)) path.push_back(s.substr(index,i-index+1));
			else continue;
			backtracking(s,i+1);
			path.pop_back();
		}
	}
	vector<vector<string>> partition(string s) {
		backtracking(s,0);
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/07-回溯/Js/02-分割/LeetCode_0131.js
//131. 分割回文串
const isPalindrome=function(s,start,end){
  while(start<end){
    if(s[start]!=s[end]) return false;
    start++;
    end--;
  }
  return true;
}
const partition = function(s) {
  const path=new Array();
  const res=new Array();
  const backtracking=function(str,index){
    if(index>=str.length){
      res.push(Array.from(path));
      return;
    }
    for(let i=index;i<str.length;i++){
      if(isPalindrome(str,index,i)) path.push(str.slice(index,i+1));
      else continue;
      backtracking(str,i+1);
      path.pop();
    }
  }
  backtracking(s,0);
  return res;
};